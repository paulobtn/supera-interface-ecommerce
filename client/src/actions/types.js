// Adiciona item ao carrinho de compras
export const ADD_TO_CART = "add_to_cart";

