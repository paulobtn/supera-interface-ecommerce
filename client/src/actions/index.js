import { ADD_TO_CART } from "./types";

/* Adiciona produto ao carrinho de compras e sua quantidade.
 * Caso jÃ¡ exista, apenas soma a quantidade
 */
export const addToCart = (qtd, data) => {
  return{
    type: ADD_TO_CART,
    payload: {qtd, data}
  }
};
