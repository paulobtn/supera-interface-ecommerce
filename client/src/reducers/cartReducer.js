import { ADD_TO_CART } from '../actions/types';

const freeShippingValue = 250;
const shippingValue = 10;

//exportando para fins de teste
export const initState = { 
  shoppingCart:[], //[{quantity, item{id,name,price,score,image}}]
  quantity: 0,
  subtotal: 0,
  shipping: 0, //apenas adicionado ao total se freeShipping for falso
  freeShipping: false,
  total: 0
}

/* stateAddToCart retorna o novo estado ao adicionar o produto ao
 * carrinho */
const stateAddToCart = (state, action) => {

  /* dado e quantidade do payload para não termos que escrever
   * action.payload varias vezes */
  let {data, quantity} = action.payload;

  /* faz uma deep copy do state, note que apenas 
   * newState = {...state}
   * iria copiar a referência de shoppingCart e não queremos
   * alterar state, apenas retornar um novo
   */
  let newState = JSON.parse(JSON.stringify(state));

  // verifica se o produto já existe
  let product = newState.shoppingCart.find( (product) => {
    return product.data.id === data.id
  });
  
  // Adiciona produto ao carrinho ou incrementa quantidade
  if(product){
    newState
      .shoppingCart[newState.shoppingCart.indexOf(product)]
      .quantity += quantity;
  } else{
    newState.shoppingCart = newState.shoppingCart.concat(action.payload);
  }
    
  // atualiza quantidade e subtotal
  newState.quantity += quantity;
  newState.subtotal += data.price * quantity;

  // atualiza fretes
  newState.shipping += quantity*shippingValue;
  if(newState.subtotal > 250){
    newState.freeShipping = true;
  }

  // atualiza total
  newState.total = newState.subtotal;
  if(!newState.freeShipping){
    newState.total += newState.shipping;
  }
  return newState;
}

const cartReducer = (state = initState, action) => {

    switch(action.type){
        case ADD_TO_CART:
          return stateAddToCart(state, action);
        default:
            return state;
    }
}

export default cartReducer;
