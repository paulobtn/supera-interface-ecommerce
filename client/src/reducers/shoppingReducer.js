import {FETCH_GAMES, SET_ERROR} from '../actions/types';

const initState = {
    items: [],
    shoppingCart:[],
    error: null
}

const shoppingReducer = (state = initState, action) => {

    switch(action.type){
        case FETCH_GAMES:
        /* Adiciona todos os jogos no campo items do 
         * redux store */
            return {
              ...state,
              error: null,
              items: action.payload
            }
        case SET_ERROR:
          return {
            ...state,
            error: action.payload
          }
        default:
            return state;
    }
}

export default shoppingReducer;
