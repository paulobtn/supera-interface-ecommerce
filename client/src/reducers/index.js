import { combineReducers } from "redux";
import shoppingReducer from "./shoppingReducer";

export default combineReducers({
  shopping: shoppingReducer
});
