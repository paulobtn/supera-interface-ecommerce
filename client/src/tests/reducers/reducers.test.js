import configureMockStore from 'redux-mock-store';

import shoppingReducer from '../../reducers/shoppingReducer';
import {initState} from '../../reducers/shoppingReducer';
import { FETCH_GAMES } from "../../actions/types";

let games = [{
  "id": 201,
  "name": "Call Of Duty Infinite Warfare",
  "price": 49.99,
  "score": 80,
  "image": "call-of-duty-infinite-warfare.png"
},
{
  "id": 102,
  "name": "The Witcher III Wild Hunt",
  "price": 119.5,
  "score": 250,
  "image": "the-witcher-iii-wild-hunt.png"
}];

describe('Testa reducer', () => {
  
  describe('shoppingReducer', () => {
    it('returnando o estado inicial', () => {
      expect(shoppingReducer(undefined, {})).toMatchSnapshot();
    });

    it('retornando o estado com os jogos carregados', () => {

      expect(
        shoppingReducer(initState, {
          type: FETCH_GAMES,
          payload: games
        })
      ).toMatchSnapshot();
    })
  });
});
