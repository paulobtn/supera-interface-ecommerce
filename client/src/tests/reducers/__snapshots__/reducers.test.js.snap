// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testa reducer shoppingReducer retornando o estado com os jogos carregados 1`] = `
Object {
  "error": null,
  "items": Array [
    Object {
      "id": 201,
      "image": "call-of-duty-infinite-warfare.png",
      "name": "Call Of Duty Infinite Warfare",
      "price": 49.99,
      "score": 80,
    },
    Object {
      "id": 102,
      "image": "the-witcher-iii-wild-hunt.png",
      "name": "The Witcher III Wild Hunt",
      "price": 119.5,
      "score": 250,
    },
  ],
  "shoppingCart": Array [],
}
`;

exports[`Testa reducer shoppingReducer returnando o estado inicial 1`] = `
Object {
  "error": null,
  "items": Array [],
  "shoppingCart": Array [],
}
`;
